File: step_04_binary_search/4.2_bs_on_answers/P003_koko_eating_bananas/solution.py
"""
Problem: Koko Eating Bananas (LeetCode #875)
Difficulty: MEDIUM | XP: 25

Koko can eat bananas at speed k per hour. Each pile takes ceil(pile/k) hours.
Given piles array and h hours, find minimum k so she can eat all within h hours.
"""
from typing import List
import math


# ============================================================
# APPROACH 1: BRUTE FORCE (Linear Scan)
# Time: O(max(piles) * n) | Space: O(1)
# ============================================================
def brute_force(piles: List[int], h: int) -> int:
    """Try every k from 1 to max(piles). Return the first k that works."""

    def total_hours(k: int) -> int:
        return sum(math.ceil(p / k) for p in piles)

    for k in range(1, max(piles) + 1):
        if total_hours(k) <= h:
            return k
    return max(piles)


# ============================================================
# APPROACH 2: OPTIMAL (Binary Search on Answer)
# Time: O(n * log(max(piles))) | Space: O(1)
# ============================================================
def optimal(piles: List[int], h: int) -> int:
    """Binary search on k in [1, max(piles)]. Check feasibility at each mid."""

    def total_hours(k: int) -> int:
        return sum(math.ceil(p / k) for p in piles)

    lo, hi = 1, max(piles)
    ans = hi
    while lo <= hi:
        mid = lo + (hi - lo) // 2
        if total_hours(mid) <= h:
            ans = mid       # feasible, try smaller k
            hi = mid - 1
        else:
            lo = mid + 1    # too slow, need faster eating
    return ans


# ============================================================
# APPROACH 3: BEST (Binary Search + Early Termination)
# Time: O(n * log(max(piles))) | Space: O(1)
# ============================================================
def best(piles: List[int], h: int) -> int:
    """Binary search with early termination in feasibility check."""

    def can_finish(k: int) -> bool:
        hours = 0
        for p in piles:
            hours += (p + k - 1) // k  # ceil without float
            if hours > h:              # early termination
                return False
        return True

    lo, hi = 1, max(piles)
    while lo < hi:
        mid = lo + (hi - lo) // 2
        if can_finish(mid):
            hi = mid        # feasible, try smaller
        else:
            lo = mid + 1    # not feasible, need larger k
    return lo


# ============================================================
# DRIVER
# ============================================================
if __name__ == "__main__":
    print("=== Koko Eating Bananas ===\n")

    test_cases = [
        ([3, 6, 7, 11], 8, 4),
        ([30, 11, 23, 4, 20], 5, 30),
        ([30, 11, 23, 4, 20], 6, 23),
        ([1], 1, 1),
        ([1000000000], 2, 500000000),
    ]

    for piles, h, expected in test_cases:
        b = brute_force(piles, h)
        o = optimal(piles, h)
        n = best(piles, h)
        status = "PASS" if b == expected and o == expected and n == expected else "FAIL"
        print(f"Input: piles={piles}, h={h}")
        print(f"  Brute:    {b}")
        print(f"  Optimal:  {o}")
        print(f"  Best:     {n}")
        print(f"  Expected: {expected}  [{status}]\n")

File: step_04_binary_search/4.2_bs_on_answers/P003_koko_eating_bananas/Solution.java
/**
 * Problem: Koko Eating Bananas (LeetCode #875)
 * Difficulty: MEDIUM | XP: 25
 *
 * Koko can eat bananas at speed k per hour. Each pile takes ceil(pile/k) hours.
 * Given piles array and h hours, find minimum k so she can eat all within h hours.
 *
 * @author DSA_Nova
 */

// ============================================================
// Approach 1: Brute Force (Linear Scan)
// Time: O(max(piles) * n) | Space: O(1)
// ============================================================
class BruteForce {
    public static int minEatingSpeed(int[] piles, int h) {
        int maxPile = 0;
        for (int p : piles) maxPile = Math.max(maxPile, p);

        for (int k = 1; k <= maxPile; k++) {
            long hours = 0;
            for (int p : piles) {
                hours += (p + k - 1) / k; // ceil(p / k)
            }
            if (hours <= h) return k;
        }
        return maxPile;
    }
}

// ============================================================
// Approach 2: Optimal (Binary Search on Answer)
// Time: O(n * log(max(piles))) | Space: O(1)
// ============================================================
class Optimal {
    public static int minEatingSpeed(int[] piles, int h) {
        int maxPile = 0;
        for (int p : piles) maxPile = Math.max(maxPile, p);

        int lo = 1, hi = maxPile, ans = maxPile;
        while (lo <= hi) {
            int mid = lo + (hi - lo) / 2;
            long hours = 0;
            for (int p : piles) {
                hours += (p + mid - 1) / mid;
            }
            if (hours <= h) {
                ans = mid;
                hi = mid - 1;
            } else {
                lo = mid + 1;
            }
        }
        return ans;
    }
}

// ============================================================
// Approach 3: Best (Binary Search + Early Termination)
// Time: O(n * log(max(piles))) | Space: O(1)
// ============================================================
class Best {
    public static int minEatingSpeed(int[] piles, int h) {
        int maxPile = 0;
        for (int p : piles) maxPile = Math.max(maxPile, p);

        int lo = 1, hi = maxPile;
        while (lo < hi) {
            int mid = lo + (hi - lo) / 2;
            if (canFinish(piles, mid, h)) {
                hi = mid;
            } else {
                lo = mid + 1;
            }
        }
        return lo;
    }

    private static boolean canFinish(int[] piles, int k, int h) {
        long hours = 0;
        for (int p : piles) {
            hours += (p + k - 1) / k;
            if (hours > h) return false; // early termination
        }
        return true;
    }
}

// Main driver
public class Solution {
    public static void main(String[] args) {
        System.out.println("=== Koko Eating Bananas ===\n");

        int[][] pilesArr = {
            {3, 6, 7, 11},
            {30, 11, 23, 4, 20},
            {30, 11, 23, 4, 20},
            {1},
            {1000000000}
        };
        int[] hArr = {8, 5, 6, 1, 2};
        int[] expected = {4, 30, 23, 1, 500000000};

        for (int t = 0; t < pilesArr.length; t++) {
            int b = BruteForce.minEatingSpeed(pilesArr[t], hArr[t]);
            int o = Optimal.minEatingSpeed(pilesArr[t], hArr[t]);
            int n = Best.minEatingSpeed(pilesArr[t], hArr[t]);
            boolean pass = (b == expected[t] && o == expected[t] && n == expected[t]);

            System.out.println("Input:    piles=" + java.util.Arrays.toString(pilesArr[t]) + ", h=" + hArr[t]);
            System.out.println("Brute:    " + b);
            System.out.println("Optimal:  " + o);
            System.out.println("Best:     " + n);
            System.out.println("Expected: " + expected[t]);
            System.out.println("Pass:     " + pass);
            System.out.println();
        }
    }
}
